nav2_smac_planner/src/a_star.cpp

```diff
@@ -200,14 +210,54 @@ void AStarAlgorithm<NodeT>::setGoal(
   const unsigned int & my,
   const unsigned int & dim_3)
 {
-  _goal = addToGraph(NodeT::getIndex(mx, my, dim_3));
-
-  typename NodeT::Coordinates goal_coords(
-    static_cast<float>(mx),
-    static_cast<float>(my),
-    static_cast<float>(dim_3));
+  _goals.clear();
+  _goalsSet.clear();
+  CoordinateVector goals_coordinates;
+  unsigned int num_bins = NodeT::motion_table.num_angle_quantization;
+  unsigned int dim_3_half_bin = 0;
+  switch (_current_goal_heading_mode) {
+    case GoalHeadingMode::DEFAULT:
+      _goals.push_back(addToGraph(NodeT::getIndex(mx, my, dim_3)));
+      goals_coordinates.push_back(
+        typename NodeT::Coordinates(
+          static_cast<float>(mx),
+          static_cast<float>(my),
+          static_cast<float>(dim_3)));
+      break;
+    case GoalHeadingMode::BIDIRECTIONAL:
+      // Add two goals, one for each direction
+      _goals.push_back(addToGraph(NodeT::getIndex(mx, my, dim_3)));
+      // 180 degrees
+      dim_3_half_bin = (dim_3 + (num_bins / 2)) % num_bins;
+      _goals.push_back(addToGraph(NodeT::getIndex(mx, my, dim_3_half_bin)));
+      goals_coordinates.push_back(
+        typename NodeT::Coordinates(
+          static_cast<float>(mx),
+          static_cast<float>(my),
+          static_cast<float>(dim_3)));
+      goals_coordinates.push_back(
+        typename NodeT::Coordinates(
+          static_cast<float>(mx),
+          static_cast<float>(my),
+          static_cast<float>(dim_3_half_bin)));
+      break;
+    case GoalHeadingMode::ALL_DIRECTION:
+      // Add all possible headings as goals
+      for (unsigned int i = 0; i < num_bins; ++i) {
+        _goals.push_back(addToGraph(NodeT::getIndex(mx, my, i)));
+        goals_coordinates.push_back(
+          typename NodeT::Coordinates(
+            static_cast<float>(mx),
+            static_cast<float>(my),
+            static_cast<float>(i)));
+      }
+      break;
+    case GoalHeadingMode::UNKNOWN:
+      throw std::runtime_error("Goal heading is UNKNOWN.");
+      break;
+  }
 
-  if (!_search_info.cache_obstacle_heuristic || goal_coords != _goal_coordinates) {
+  if (!_search_info.cache_obstacle_heuristic || goals_coordinates != _goals_coordinates) {
```

Comment:
Does really matter if the yaw matches if the goal is to reset the obstacle heurtistic which does not depend on the information on the yaw?

Reply:
I am not quite sure I understand. 

The obstacle critic should only be reset when the map coords(costmap coords) are changed. Since all the goals should have the same x,y you can just check the initial of both sets to determine if the obstacle heuristic needs to be recalculated.  

It doesn't matter if the yaw matches, but it does matter if there is a mismatch because we will end up recomputing the obstacle heuristic for changes in yaw which has no affect on the obstacle heuristic.